tools/validate-image/valImage.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2019 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"knative.dev/test-infra/tools/monitoring/mail"
+	"knative.dev/test-infra/tools/monitoring/subscriber"
+)
+
+var (
+	// monitoringSubs is a list of subscriptions to subscribe for image vulnerabilities
+	monitoringSubs = [...]string{
+		"sub-container-analysis-notes-v1beta1",
+		"sub-container-analysis-occurrences-v1beta1",
+	}
+	recipients = []string{"knative-productivity-dev@googlegroups.com"}
+
+	// alertFreq is the minimum wait time before sending another image vulnerability alert
+	alertFreq = 24 * time.Hour
+
+	// Cache the last alert time in memory to prevent multiple image
+	// vulnerability alerts sent in a short duration of time.
+	lastSent = time.Time{}
+)
+
+// Client holds resources for monitoring image vulnerabilities
+type Client struct {
+	subClients []*subscriber.Client
+	mailClient *mail.Config
+}
+
+// NewValidateImageClient initialize all the resources for monitoring image vulnerabilities
+func NewValidateImageClient(mconfig *mail.Config) (*Client, error) {
+	var subClients = make([]*subscriber.Client, 0)
+
+	for _, sub := range monitoringSubs {
+		log.Printf("Appending sub: %v\n", sub)
+		subc, err := subscriber.NewSubscriberClient(sub)
+		if err != nil {
+			return nil, err
+		}
+		subClients = append(subClients, subc)
+		log.Printf("subclients: %v\n", subClients)
+	}
+
+	return &Client{
+		subClients: subClients,
+		mailClient: mconfig,
+	}, nil
+}
+
+// Run start a background process that listens to the message
+func (c *Client) Run() {
+	log.Println("Starting image vulnerabilities monitoring")
+	for _, sub := range c.subClients {
+		c.listen(sub)
+	}
+}
+
+func (c *Client) listen(subClient *subscriber.Client) {
+	go func() {
+		err := subClient.Receive(context.Background(), func(ctx context.Context, msg *pubsub.Message) {
+			log.Printf("Message: %v\n", string(msg.Data))
+			log.Printf("Pubsub Message: %v\n", msg)
```

Comment:
whats the point of logging both? This will log the data as well?

Reply:
The one logged with `string(msg.Data)` prints data in a human-readable format, while printing the message directly from msg prints the message as base64encoded string. 

I'd like to keep the pubsub message for debugging purposes in case there's additional attributes that gets added to the image vulnerabilities message. 

```
# log.Printf("Message: %v\n", string(msg.Data))
Message Data: {"name":"projects/joyceyu-test/occurrences/ce3b5a33-05ed-40c2-80f2-821d38426171","kind":"DISCOVERY","notificationTime":"2019-09-12T18:04:36.630196Z"}

# log.Printf("Pubsub Message: %v\n", msg)
Pubsub Message: {
        "ID": "719306958870977",
        "Data": "eyJuYW1lIjoicHJvamVjdHMvam95Y2V5dS10ZXN0L29jY3VycmVuY2VzL2NlM2I1YTMzLTA1ZWQtNDBjMi04MGYyLTgyMWQzODQyNjE3MSIsImtpbmQiOiJESVNDT1ZFUlkiLCJub3RpZmljYXRpb25UaW1lIjoiMjAxOS0wOS0xMlQxODowNDozNi42MzAxOTZaIn0=",
        "Attributes": null,
        "PublishTime": "2019-09-12T18:04:37.058Z"
}
```